corehq/apps/userreports/app_manager/data_source_meta.py

```diff
@@ -41,3 +45,112 @@ def make_form_data_source_filter(xmlns):
         },
         "property_value": xmlns,
     }
+
+
+def get_app_data_source_meta(domain, data_source_type, data_source_id):
+    """
+    Get an AppDataSourceMeta object based on the expected domain, source type and ID
+    :param domain: the domain
+    :param data_source_type: must be "form" or "case"
+    :param data_source_id: for forms - the unique form ID, for cases the case type
+    :return: an AppDataSourceMeta object corresponding to the data source type and ID
+    """
+    assert data_source_type in DATA_SOURCE_TYPE_VALUES
+    if data_source_type == DATA_SOURCE_TYPE_CASE:
+        return CaseDataSourceMeta(domain, data_source_type, data_source_id)
+    else:
+        return FormDataSourceMeta(domain, data_source_type, data_source_id)
+
+
+class AppDataSourceMeta(six.with_metaclass(ABCMeta, object)):
+    """
+    Utility base class for interacting with forms/cases inside an app and data sources
+    """
+
+    def __init__(self, domain, data_source_type, data_source_id):
+        self.domain = domain
+        self.data_source_type = data_source_type
+        self.data_source_id = data_source_id
+
+    def get_doc_type(self):
+        return get_data_source_doc_type(self.data_source_type)
+
+    @abstractmethod
+    def get_filter(self):
+        return make_case_data_source_filter
```

Comment:
Was this left on purpose? I think this might fail if you ever call `super` in the subclasses.

Reply:
good catch. https://github.com/dimagi/commcare-hq/pull/19987/commits/f7d74b1a06efbeb7c8e87f8d6862f0adba0d8952